### ISSUE TITLE
Unknown field with tag 65428 (0xff94) encountered  error with more than 4.1GB ndpi

### ISSUE THREAD

**Issue by owenustc** [libvips/libvips#1206]
Hi, 
    I had met a error, when i tried to convert a big file with ndpi format, about 4.1G.  My command as follow:

```
$ vips dzsave  S100233-HE12.ndpi  S100233-HE12.ndpi.map  --compression -1  --background 0 --centre --layout google
```

The error information as follow:

```
(vips:27237): VIPS-WARNING **: ASCII value for tag "Tag 65459" does not end in null byte. Forcing it to be null

(vips:27237): VIPS-WARNING **: Unknown field with tag 65420 (0xff8c) encountered

(vips:27237): VIPS-WARNING **: Unknown field with tag 65421 (0xff8d) encountered

(vips:27237): VIPS-WARNING **: Unknown field with tag 65422 (0xff8e) encountered

(vips:27237): VIPS-WARNING **: Unknown field with tag 65423 (0xff8f) encountered

(vips:27237): VIPS-WARNING **: Unknown field with tag 65424 (0xff90) encountered

(vips:27237): VIPS-WARNING **: Unknown field with tag 65425 (0xff91) encountered

(vips:27237): VIPS-WARNING **: Unknown field with tag 65426 (0xff92) encountered

(vips:27237): VIPS-WARNING **: Unknown field with tag 65427 (0xff93) encountered

(vips:27237): VIPS-WARNING **: Unknown field with tag 65428 (0xff94) encountered

(vips:27237): VIPS-WARNING **: Unknown field with tag 65433 (0xff99) encountered

(vips:27237): VIPS-WARNING **: Unknown field with tag 65439 (0xff9f) encountered

(vips:27237): VIPS-WARNING **: Unknown field with tag 65441 (0xffa1) encountered

(vips:27237): VIPS-WARNING **: Unknown field with tag 65442 (0xffa2) encountered

(vips:27237): VIPS-WARNING **: Unknown field with tag 65443 (0xffa3) encountered

(vips:27237): VIPS-WARNING **: Unknown field with tag 65444 (0xffa4) encountered

(vips:27237): VIPS-WARNING **: Unknown field with tag 65445 (0xffa5) encountered

(vips:27237): VIPS-WARNING **: Unknown field with tag 65446 (0xffa6) encountered

(vips:27237): VIPS-WARNING **: Unknown field with tag 65455 (0xffaf) encountered
(vips:27237): VIPS-WARNING **: Unknown field with tag 65456 (0xffb0) encountered

(vips:27237): VIPS-WARNING **: Unknown field with tag 65457 (0xffb1) encountered

(vips:27237): VIPS-WARNING **: Unknown field with tag 65458 (0xffb2) encountered

(vips:27237): VIPS-WARNING **: Unknown field with tag 65459 (0xffb3) encountered

(vips:27237): VIPS-WARNING **: error in tile 0 x 0
TIFFFetchDirectory: Sanity check on directory count failed, this is probably not a valid IFD offset
TIFFReadDirectory: Failed to read directory at offset 1195726
TIFFFetchDirectory: Sanity check on directory count failed, this is probably not a valid IFD offset
TIFFReadDirectory: Failed to read directory at offset 1195726
JPEGLib: Empty JPEG image (DNL not supported)
```

And the version of my vips is 8.7.

**Comment by jcupitt** [comment]
Hello @owenustc,

It looks like your libvips has been built without openslide support. I see:

```
john@kiwi:~$ vips openslideload
load file with OpenSlide
usage:
   openslideload filename out [--option-name option-value ...]
where:
   filename     - Filename to load from, input gchararray
   out          - Output image, output VipsImage
optional arguments:
.. etc.
```

Without openslide, `.ndpi` files just look like plain tiff files, but with extra incomprehensible fields. 

What platform are you using, and where did you get the libvips binary from?

**Comment by owenustc** [comment]
Hi， I build libvips with openslide support, which as follow:
nodeip-dev115:/home/owen/convert# vips openslideload
load file with OpenSlide
usage:
   openslideload filename out [--option-name option-value ...]
where:
   filename     - Filename to load from, input gchararray
   out          - Output image, output VipsImage
optional arguments:
   flags        - Flags for this file, output VipsForeignFlags
   memory       - Force open via memory, input gboolean
                        default: false
   access       - Required access pattern for this file, input VipsAccess
                        default: random
                        allowed: random, sequential, sequential-unbuffered
   level        - Load this level from the file, input gint
                        default: 0
                        min: 0, max: 100000
   autocrop     - Crop to image bounds, input gboolean
                        default: false
   fail         - Fail on first error, input gboolean
                        default: false
   associated   - Load this associated image, input gchararray
And I detected the problem could be from openslide configuration, which i do two test code as follow:

root@nodeip-dev115:/home/owen/convert# cat check.py
import sys
import re,os
import openslide
import time
from openslide import OpenSlide, OpenSlideError
from openslide.deepzoom import DeepZoomGenerator
from PIL import Image
def check(filename):
    Image.MAX_IMAGE_PIXELS = 1024*1024*1024*10*8
    org=openslide.open_slide(filename)
    print (org.dimensions) 
The result of 4.1GB ndpi is right, which as follow:
python check.py S100233-HE12.ndpi
(168960, 87552)
And if I disable the  configuration of Image.MAX_IMAGE_PIXELS,  it will generated the result as follow:
t@nodeip-dev115:/home/owen/convert# python check.py  S100233-HE12.ndpi
/usr/lib/python2.7/dist-packages/PIL/Image.py:2224: DecompressionBombWarning: Image size (14792785920 pixels) exceeds limit of 89478485 pixels, could be decompression bomb DOS attack.
  DecompressionBombWarning)
(168960, 87552)

So I guess this configuration could be set by some parameter of the code of vips.

**Comment by jcupitt** [comment]
Perhaps your openslide does not have bigtiff support? 

What version of libtiff is openslide built against? It needs to be version 4 or later for files larger than 4 GB.

To test it, try:

```python
import sys
import pyvips

x = pyvips.Image.openslideload(sys.argv[1])
print(x.width, x.height)
```

Then you should see something like:

```
$ python check.py 2013_09_20_29.ndpi 
(118784, 102400)
```

**Comment by jcupitt** [comment]
That file is 4.1GB, by the way:

```
-rw-r--r-- 1 john john 4101070956 May  7  2015  2013_09_20_29.ndpi
```

So it should work for you.

**Comment by owenustc** [comment]
My libtiff  is libtiff5.  And I meet the error with your code, which error as follow:
ost recent call last):
  File "check.py", line 19, in <module>
    check2(sys.argv[1])
  File "check.py", line 16, in check2
    x = pyvips.Image.openslideload(filename)
  File "/usr/local/lib/python2.7/dist-packages/pyvips/vimage.py", line 128, in call_function
    return pyvips.Operation.call(name, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/pyvips/voperation.py", line 189, in call
    raise Error('unable to call {0}'.format(operation_name))
pyvips.error.Error: unable to call openslideload
  openslide2vips: unsupported slide format

**Comment by owenustc** [comment]
support 8bpp PNG quantisation:          no
  (requires libimagequant)
file import/export with libtiff:        yes (pkg-config libtiff-4)
file import/export with giflib:         yes (found by search)
file import/export with libjpeg:        yes (pkg-config)
image pyramid export:                   yes

My configuration

**Comment by owenustc** [comment]
ome/owen/tool/libvips-8.7.3/tools/vips dzsave  S101600-HE5-6.ndpi S101600-HE5-6.ndpi.map  --compression -1  --background 0 --centre --layout google

(lt-vips:55590): VIPS-WARNING **: Unknown field with tag 65420 (0xff8c) encountered

(lt-vips:55590): VIPS-WARNING **: Unknown field with tag 65421 (0xff8d) encountered

(lt-vips:55590): VIPS-WARNING **: Unknown field with tag 65422 (0xff8e) encountered

(lt-vips:55590): VIPS-WARNING **: Unknown field with tag 65423 (0xff8f) encountered

**Comment by jcupitt** [comment]
Could you perhaps have several versions of libvips, openslide or libtiff installed? What platform are you using? Some platforms will have several versions of libtiff installed, and your libvips might be linked to the wrong one.

As I said, I can work with >4GB ndpi images here, so I think it must be an issue with the way your binary has been made.

The best way to track this down would be to run in a debugger and verify which library is being used. For example, I built openslide with debugging, then tried:

```
$ gdb python
Reading symbols from python...(no debugging symbols found)...done.
(gdb) break openslide_open
Function "openslide_open" not defined.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (openslide_open) pending.
(gdb) run ~/try/check.py 2013_09_20_29.ndpi 
Starting program: /usr/bin/python ~/try/check.py 2013_09_20_29.ndpi
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Breakpoint 1, openslide_open (filename=0x555555eca5d0 "2013_09_20_29.ndpi")
    at src/openslide.c:237
237	openslide_t *openslide_open(const char *filename) {
```

You could continue to the ndpi reader:

```
(gdb) break hamamatsu_ndpi_open
Breakpoint 2 at 0x7ffff1ff45d6: file src/openslide-vendor-hamamatsu.c, line 2233.
(gdb) c
Continuing.

Breakpoint 2, hamamatsu_ndpi_open (osr=0x555555ba6c00, 
    filename=0x555555eca5d0 "2013_09_20_29.ndpi", tl=0x555555cf3040, 
    quickhash1=0x555555f9ab00, err=0x7fffffffcc68)
    at src/openslide-vendor-hamamatsu.c:2233
2233	                                GError **err) {
```

Then try to see where it fails.

**Comment by owenustc** [comment]
oot@nodeip-dev115:/home/owen/AI/convert# ldd /usr/local/bin/vips | more
        linux-vdso.so.1 =>  (0x00007ffe3f4cf000)
        libvips.so.42 => /usr/local/lib/libvips.so.42 (0x00002b88623da000)
        libgobject-2.0.so.0 => /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0 (0x00002b88629af000)
        libglib-2.0.so.0 => /lib/x86_64-linux-gnu/libglib-2.0.so.0 (0x00002b8862c02000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00002b8862f13000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00002b8863130000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00002b88634fa000)
        libMagickCore-6.Q16.so.2 => /usr/lib/x86_64-linux-gnu/libMagickCore-6.Q16.so.2 (0x00002b8863714000)
        libpng12.so.0 => /usr/lib/x86_64-linux-gnu/libpng12.so.0 (0x00002b8863bc8000)
        libtiff.so.5 => /usr/lib/x86_64-linux-gnu/libtiff.so.5 (0x00002b8863ded000) ///Here 
        libjpeg.so.8 => /usr/lib/x86_64-linux-gnu/libjpeg.so.8 (0x00002b8864061000)
        libgmodule-2.0.so.0 => /usr/lib/x86_64-linux-gnu/libgmodule-2.0.so.0 (0x00002b88642ba000)
        libexpat.so.1 => /lib/x86_64-linux-gnu/libexpat.so.1 (0x00002b88644be000)
        libgsf-1.so.114 => /usr/lib/x86_64-linux-gnu/libgsf-1.so.114 (0x00002b88646e7000)
        libfftw3.so.3 => /usr/lib/x86_64-linux-gnu/libfftw3.so.3 (0x00002b886492c000)
        liborc-0.4.so.0 => /usr/lib/x86_64-linux-gnu/liborc-0.4.so.0 (0x00002b8864d2a000)
        liblcms2.so.2 => /usr/lib/x86_64-linux-gnu/liblcms2.so.2 (0x00002b8864faa000)
        libgif.so.7 => /usr/lib/x86_64-linux-gnu/libgif.so.7 (0x00002b8865201000)
        librsvg-2.so.2 => /usr/lib/x86_64-linux-gnu/librsvg-2.so.2 (0x00002b886540a000)
        libcairo.so.2 => /usr/lib/x86_64-linux-gnu/libcairo.so.2 (0x00002b8865641000)
        libIlmImf-2_2.so.22 => /usr/lib/x86_64-linux-gnu/libIlmImf-2_2.so.22 (0x00002b8865955000)
        libopenslide.so.0 => /usr/lib/x86_64-linux-gnu/libopenslide.so.0 (0x00002b8865

**Comment by owenustc** [comment]
S101600-HE5-6.ndpi
<https://drive.google.com/file/d/1aMaCEyY3HVi8kVEXmnDadSkJhxH3FUiJ/view?usp=drive_web>
The file is 4.3G ndpi.  My tool meet  error with this file.

John Cupitt <notifications@github.com> 于2019年1月11日周五 下午9:40写道：

**Comment by jcupitt** [comment]
Thanks! I'll try with your file here.

**Comment by jcupitt** [comment]
What platform are you using? Is it Ubuntu? Which version?

I'll make you a dockerfile with a working libvips in.

**Comment by jcupitt** [comment]
Yes, your file works fine for me:

```
$ ~/try/check.py S101600-HE5-6.ndpi 
(190464, 104704)
```

So you certainly have a build issue. Tell me your platform and I can make a build file for you.

**Comment by jcupitt** [comment]
For example, here it is on Alpine:

https://github.com/jcupitt/docker-builds/tree/master/libvips-openslide-alpine

I see:

```
$ docker run --rm -it -v /home/john/pics/openslide:/data libvips-openslide-alpine bash
bash-4.4# vipsheader /data/2013_09_20_29.ndpi 
/data/2013_09_20_29.ndpi: 118784x102400 uchar, 4 bands, rgb, openslideload
```

**Comment by owenustc** [comment]
My platform is a ubuntu-based docker.

**Comment by owenustc** [comment]
Do you try this command:  vips dzsave S1818532.ndpi S1818532.ndpi.map  --compression -1  --background 0 --centre --layout google?

**Comment by owenustc** [comment]
And My CPU is x86_64.

**Comment by owenustc** [comment]
Could you build a ubuntu-based docker image with vips to work with big file?  I can receive the image of docker by owenustc@gmail.com.

**Comment by jcupitt** [comment]
I tried 18.04 and the Ubuntu libvips works fine there too.

**Comment by owenustc** [comment]
root@nodeip-dev115:/home/owen/AI/smart-augment/tests# cat /etc/issue
Ubuntu 16.04.4 LTS \n \l
Do you mean that there is libvips in  ubuntu 18.10?

**Comment by owenustc** [comment]
Could you  provide your docker image with ubuntu 18.04 to me by email?

**Comment by jcupitt** [comment]
For 18.04, there's no docker image, you just install standard packages:

```
FROM ubuntu:bionic
  
RUN apt-get update \
        && apt-get install -y
    
RUN apt-get install -y \
        libvips42 \ 
        python3-pip

RUN pip3 install pyvips
```

**Comment by jcupitt** [comment]
Yes, Ubuntu has always had a libvips package I think. I don't know if the 16.04 one has openslide support.

**Comment by owenustc** [comment]
Thanks  @jcupitt , the problem is fixed

**Comment by jcupitt** [comment]
OK, glad it's working!

**Comment by owenustc** [comment]
S105330-HE1718.ndpi
<https://drive.google.com/file/d/1C0EBm-r2aZ0ahE_n7LuufGUZYu40L-zP/view?usp=drive_web>
Hi, the file attached is my big file with 5.7GB

John Cupitt <notifications@github.com> 于2019年1月16日周三 上午2:43写道：

**Comment by jcupitt** [comment]
See https://openslide.org/#getting-help

**Comment by owenustc** [comment]
Hi @jcupitt , can you generate a big file with 5.1 GB, and try you tool to validate your guess.

**Comment by jcupitt** [comment]
No, I can't make .ndpi files. 

It works fine with S101600-HE5-6.ndpi, which is 4.3GB and over the 32-bit limit, so I think it's not a libvips or openslide problem.

**Comment by jcupitt** [comment]
We shouldn't post on a closed issue -- let's move this to #1216
